File: python/c30_3_convex_hull_problem.py
import unittest
from c29_2_points_and_lines import point_location

def andrew(points):
    if len(points)<2:
        return points
    points=sorted(points,key=lambda p:(p.real,p.imag))
    upper=points[:2]
    for i in range(2,len(points)):
        upper.append(points[i])
        while len(upper)>2 and point_location(*upper[-3:])>=0:
            del upper[-2]
    lower=points[-1:-3:-1]
    for i in range(len(points)-3,-1,-1):
        lower.append(points[i])
        while len(lower)>2 and point_location(*lower[-3:])>=0:
            del lower[-2]
    return upper[:-1]+lower[:-1]

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual(andrew([]),[])
        self.assertEqual(andrew([0]),[0])
        self.assertEqual(andrew([0,1]),[0,1])
        self.assertEqual(andrew([0,1,1+1j]),[0,1+1j,1])
        self.assertEqual(andrew([0,1,2]),[0,2])
        self.assertEqual(andrew([0+1j,0+3j,1+2j,2+3j,2+5j,3+3j,4+0j,4+1j,4+4j,5+3j,6+2j,6+4j,7+2j]),[0+1j,0+3j,2+5j,6+4j,7+2j,4+0j])

if __name__=='__main__':
    unittest.main()

File: c++/c30_3_convex_hull_problem.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c29_2_points_and_lines.h"

namespace {
	vector<P> andrew(vector<P> points) {
		if (points.size() < 2)
			return points;
		ranges::sort(points, {}, [](P p) { return make_pair(p.real(), p.imag()); });
		vector<P> upper = { points[0], points[1] };
		for (ll i = 2; i < ssize(points); ++i) {
			upper.push_back(points[i]);
			while (upper.size() > 2 && point_location(upper[upper.size() - 3], upper[upper.size() - 2], upper[upper.size() - 1]) >= 0)
				upper.erase(upper.end() - 2);
		}
		upper.pop_back();
		vector<P> lower = { points[points.size() - 1], points[points.size() - 2] };
		for (ll i = ssize(points) - 3; i >= 0; --i) {
			lower.push_back(points[i]);
			while (lower.size() > 2 && point_location(lower[lower.size() - 3], lower[lower.size() - 2], lower[lower.size() - 1]) >= 0)
				lower.erase(lower.end() - 2);
		}
		lower.pop_back();
		upper.insert(upper.end(), lower.begin(), lower.end());
		return upper;
	}
}

TEST(C303ConvexHullProblem, andrew) {
	EXPECT_EQ(andrew({}), vector<P>{});
	EXPECT_EQ(andrew({ {0,0} }), (vector<P>{ {0,0} }));
	EXPECT_EQ(andrew({ {0,0},{1,0} }), (vector<P>{ {0,0},{1,0} }));
	EXPECT_EQ(andrew({ {0,0},{1,0},{1,1} }), (vector<P>{ {0,0},{1,1},{1,0} }));
	EXPECT_EQ(andrew({ {0,0},{1,0},{2,0} }), (vector<P>{ {0,0},{2,0} }));
	EXPECT_EQ(andrew({ {0,1},{0,3},{1,2},{2,3},{2,5},{3,3},{4,0},{4,1},{4,4},{5,3},{6,2},{6,4},{7,2} }), (vector<P>{ {0,1},{0,3},{2,5},{6,4},{7,2},{4,0} }));
}
